### DISCUSSION TITLE
FINN Development in a VS Code Dev Container (on Windows using WSL 2)

### DISCUSSION THREAD

**Discussion by fpjentzsch** [Xilinx/finn#1084] [Show and tell]
Hi,

I've been working with FINN in Visual Studio Code on Windows via a WSL 2 Ubuntu distribution and Docker Desktop for a while. This works fine, but the disconnect between writing code in the WSL workspace but executing it within the FINN Docker container (via `run-docker.sh`) is not ideal.

To work directly _within_ the container, I created the [devcontainer.json](https://github.com/user-attachments/files/16070821/devcontainer.json) file I'm sharing here. It allows VS Code to spin up the FINN Docker environment as a dev container, in which your development workspace resides. This has the following benefits:
- Full IntelliSense (e.g., Pylance) with up to date dependencies
- Pytest integration
- Python debugger integration
- Ability to launch multiple terminals or Jupyter notebooks directly in VS Code
- It's even easier to work with local forks of dependencies (simply comment-in the respective lines in the devcontainer.json)

### Setup
Besides some VS Code configuration via `settings.json` and `launch.json` files, you'll need to append the following code to `Dockerfile.finn` for this to work:
```
ARG USERNAME=finn-user
ARG USER_UID=1000
ARG USER_GID=$USER_UID
RUN groupadd --gid $USER_GID $USERNAME \
    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
    && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
    && chmod 0440 /etc/sudoers.d/$USERNAME
USER $USERNAME
```
Other considerations:
- Your workspace (where finn/ is cloned) should not be = $HOME, otherwise the container will mount your entire $HOME.
- Make sure to set all environment variables required by FINN, especially those that have no default value defined in devcontainer.json.

### Feedback
Please share your ideas for improving this setup here. I'll try to maintain this post accordingly.
**Note that this is not an officially supported installation of FINN.**

**Comment by fpjentzsch** [comment]
I also tried to experiment with a custom BuildKit frontend, for example [this one](https://codeberg.org/devthefuture/dockerfile-x) to apply the patch to Dockerfile.finn by extending instead of modifying it.

Unfortunately, custom frontends like this are broken in dev containers until https://github.com/microsoft/vscode-remote-release/issues/6848 is addressed.

**Comment by kalahel** [comment]
Hello, 

I have a quick question, how do you manage FINN call to vitis_hls since it is a windows install of Vitis ?

Personally when launching `notebooks/end2end_example/bnn-pynq/cnv_end2end_example.py` I have `AssertionError: vitis_hls not found in PATH` even though I have given the xilinx installation path.
Also my `/mnt/c/Tools/Xilinx/Vitis_HLS/2024.1/bin/vitis_hls` cannot run since there is no linux folder installation.

That what I get if I try to run it : 
```shell
ERROR: Could not find 64-bit executable.
ERROR: /mnt/c/Tools/Xilinx/Vitis_HLS/2024.1/bin/unwrapped/lnx64.o/vitis_hls does not exist
```

**Reply by fpjentzsch** [reply]
Hi,

I actually install Vitis/Vivado **inside** my WSL 2 distribution so FINN can use it. If you have WSLg enabled, you can simply use the GUI-based Linux web installer. Otherwise you can execute the installer via command line.

**Reply by kalahel** [reply]
Thanks, that's very clever !

**Reply by San0519** [reply]
Hi,

I am currently working on executing FINN inside my WSL 2 with Vivado/Vitis HLS being installed on my host machine. From your reply, might I regard that as it would not be supported even the volumn of Vivado/Vitis HLS have been mounted to the Docker container, like referred as "/mnt/c/Xilinx/...".

**Reply by fpjentzsch** [reply]
No, that will not work. You'll need to install Vitis/Vivado within WSL 2.

**Reply by San0519** [reply]
Great! Thanks for your answer.

**Comment by fpjentzsch** [comment]
As of [this PR](https://github.com/eki-project/finn-plus/pull/51) FINN+ switched to a containerless Poetry-based setup. This makes most of this guide obsolete as you can run FINN directly from a Python virtual environment in VS Code connected to WSL 2.

**Reply by SanderDebets** [reply]
Hi, does that mean that the docker setup is not required for running notebooks (so doing bash ./run-docker.sh notebook is not needed)? I've been trying to run notebooks/end2end_example/cybersecurity/3-build-accelerator-with-finn.ipynb but was getting errors in the build dataflow step. This was when running in VS Code connected to WSL 2. So then I figured I should try using the docker setup. But I had some difficulty getting this to work as well.

**Reply by fpjentzsch** [reply]
Hi,
yes, in [FINN+](https://github.com/eki-project/finn-plus) Docker is not used.
FINN+ doesn't come with a built-in Jupyter server like FINN (./run-docker.sh notebook), but since it is based on a Poetry-managed Python venv, you can run Jupyter notebooks directly in VS Code with the Jupyter extension if you select that venv as the kernel. Or you can setup your own Jupyter server.

**Reply by SanderDebets** [reply]
Ah I see, Is the idea of FINN+ for it to purely be an extension/improvement of FINN, or can i expect basic functionalities to be different in FINN+ as compared to FINN?

**Reply by fpjentzsch** [reply]
Besides the containerless installation, the basic functionality should be the same, as we regularly pull in all changes from FINN.

However, we might extend or fix bugs in certain transformations such that custom build flows will need to be adapted when moving from FINN to FINN+. For example, the old ResNet-50 finn-example needs different custom streamlining steps in FINN+.

**Reply by SanderDebets** [reply]
Great, thanks for the info!
